AdyenCardScanner/Sources/CardImageParser.swift

```diff
@@ -0,0 +1,233 @@
+        colorControlsFilter?.setValue(0.2, forKey: kCIInputBrightnessKey)
+        colorControlsFilter?.setValue(1.5, forKey: kCIInputContrastKey)
+        colorControlsFilter?.setValue(1.2, forKey: kCIInputSaturationKey)
+        outputImage = colorControlsFilter?.outputImage
+
+        // Apply sharpness enhancement filter.
+        let sharpenFilter = CIFilter(name: "CISharpenLuminance")
+        sharpenFilter?.setValue(outputImage, forKey: kCIInputImageKey)
+        sharpenFilter?.setValue(0.5, forKey: kCIInputSharpnessKey)
+        outputImage = sharpenFilter?.outputImage
+
+//        outputImage = convertToGrayscale(image: outputImage!)
+
+        return outputImage
+    }
+
+    // ================== DO NOT REMOVE ==================
+
+    func processImageForTextRecognition(image: CIImage?, threshold: CGFloat) -> CIImage? {
+        guard let image else { return nil }
+        if let grayscaleImage = convertToGrayscale(image: image) {
+            return applyThreshold(image: grayscaleImage, threshold: threshold)
+        }
+        return nil
+    }
+
+    private func convertToGrayscale(image: CIImage) -> CIImage? {
+        let colorControlsFilter = CIFilter(name: "CIColorControls")
+        colorControlsFilter?.setValue(image, forKey: kCIInputImageKey)
+        colorControlsFilter?.setValue(0.0, forKey: kCIInputSaturationKey) // No color
+        colorControlsFilter?.setValue(1.0, forKey: kCIInputBrightnessKey) // Full brightness
+        colorControlsFilter?.setValue(1.0, forKey: kCIInputContrastKey) // Full contrast
+        return colorControlsFilter?.outputImage
+    }
+
+    private func applyThreshold(image: CIImage, threshold: CGFloat) -> CIImage? {
+        let colorMatrixFilter = CIFilter(name: "CIColorMatrix")
+        colorMatrixFilter?.setValue(image, forKey: kCIInputImageKey)
+
+        // Set the threshold value
+        let thresholdVector = CIVector(x: threshold, y: threshold, z: threshold, w: 1)
+        colorMatrixFilter?.setValue(thresholdVector, forKey: "inputRVector")
+        colorMatrixFilter?.setValue(thresholdVector, forKey: "inputGVector")
+        colorMatrixFilter?.setValue(thresholdVector, forKey: "inputBVector")
+
+        // Keeping alpha channel as is
+        colorMatrixFilter?.setValue(CIVector(x: 0, y: 0, z: 0, w: 1), forKey: "inputAVector")
+
+        // Biasing the output so that it’s either black or white
+        let biasVector = CIVector(x: -threshold, y: -threshold, z: -threshold, w: 0)
+        colorMatrixFilter?.setValue(biasVector, forKey: "inputBiasVector")
+
+        return colorMatrixFilter?.outputImage
+    }
+
+    // ==================================================
+
+    private func extractCardNumber(from textObservations: [VNRecognizedTextObservation]) -> String? {
+        for observation in textObservations {
+            let candidates = observation.topCandidates(Constants.topCandidates)
+            guard let topCandidate = candidates.first,
```

Comment:
How about moving this logic to a new function to get rid of `continue`? May also help this function look simple with like 2-3 lines.

Reply:
Do you suggest filtering the candidates first, and then traversing the list?

Reply:
A function inside the for loop so we call 
```
for observation in textObservations {
    let candidates = observation.topCandidates(Constants.topCandidates)
    findCardNumber(from candidates)
}
```
or move candidate line into the function too etc

Reply:
Refactor the method. No for loop is present